// plugin-welcome/src/api/models/EntSubjectType.ts
/* tslint:disable */
/* eslint-disable */
/**
 * SUT SA Project API Course Item
 * This is a sample server for SUT SE 2563
 *
 * The version of the OpenAPI document: 1.0
 * Contact: support@swagger.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    EntSubjectTypeEdges,
    EntSubjectTypeEdgesFromJSON,
    EntSubjectTypeEdgesFromJSONTyped,
    EntSubjectTypeEdgesToJSON,
} from './';

/**
 * 
 * @export
 * @interface EntSubjectType
 */
export interface EntSubjectType {
    /**
     * 
     * @type {EntSubjectTypeEdges}
     * @memberof EntSubjectType
     */
    edges?: EntSubjectTypeEdges;
    /**
     * ID of the ent.
     * @type {number}
     * @memberof EntSubjectType
     */
    id?: number;
    /**
     * TypeName holds the value of the "type_name" field.
     * @type {string}
     * @memberof EntSubjectType
     */
    typeName?: string;
}

export function EntSubjectTypeFromJSON(json: any): EntSubjectType {
    return EntSubjectTypeFromJSONTyped(json, false);
}

export function EntSubjectTypeFromJSONTyped(json: any, ignoreDiscriminator: boolean): EntSubjectType {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'edges': !exists(json, 'edges') ? undefined : EntSubjectTypeEdgesFromJSON(json['edges']),
        'id': !exists(json, 'id') ? undefined : json['id'],
        'typeName': !exists(json, 'type_name') ? undefined : json['type_name'],
    };
}

export function EntSubjectTypeToJSON(value?: EntSubjectType | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'edges': EntSubjectTypeEdgesToJSON(value.edges),
        'id': value.id,
        'type_name': value.typeName,
    };
}
