// plugin-welcome/src/api/models/EntSubjectEdges.ts
/* tslint:disable */
/* eslint-disable */
/**
 * SUT SA Project API Course Item
 * This is a sample server for SUT SE 2563
 *
 * The version of the OpenAPI document: 1.0
 * Contact: support@swagger.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    EntCourseItem,
    EntCourseItemFromJSON,
    EntCourseItemFromJSONTyped,
    EntCourseItemToJSON,
    EntTeacher,
    EntTeacherFromJSON,
    EntTeacherFromJSONTyped,
    EntTeacherToJSON,
} from './';

/**
 * 
 * @export
 * @interface EntSubjectEdges
 */
export interface EntSubjectEdges {
    /**
     * CourseItems holds the value of the course_items edge.
     * @type {Array<EntCourseItem>}
     * @memberof EntSubjectEdges
     */
    courseItems?: Array<EntCourseItem>;
    /**
     * 
     * @type {EntTeacher}
     * @memberof EntSubjectEdges
     */
    owner?: EntTeacher;
}

export function EntSubjectEdgesFromJSON(json: any): EntSubjectEdges {
    return EntSubjectEdgesFromJSONTyped(json, false);
}

export function EntSubjectEdgesFromJSONTyped(json: any, ignoreDiscriminator: boolean): EntSubjectEdges {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'courseItems': !exists(json, 'courseItems') ? undefined : ((json['courseItems'] as Array<any>).map(EntCourseItemFromJSON)),
        'owner': !exists(json, 'Owner') ? undefined : EntTeacherFromJSON(json['Owner']),
    };
}

export function EntSubjectEdgesToJSON(value?: EntSubjectEdges | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'courseItems': value.courseItems === undefined ? undefined : ((value.courseItems as Array<any>).map(EntCourseItemToJSON)),
        'owner': EntTeacherToJSON(value.owner),
    };
}
